src/index.ts

```diff
@@ -132,53 +132,113 @@ async function handle_get(request: Request, bucket: R2Bucket): Promise<Response>
+		};
+		switch (range_type.type) {
+			case 0:
+				return headers;
+			case 1: {
+				const { start, end } = range_type;
+				return {
+					...headers,
+					// When end is not specified, set end to the maximum index of the object
+					'Content-Range': `bytes ${start}-${end || object.size - 1}/${object.size}`,
+				};
+			}
+			case 2: {
+				headers['Content-Type'] = `multipart/byteranges; boundary=${range_type.boundary}`;
+				return headers;
+			}
+		}
+	};
+
+	const rangeHeader = request.headers.get('Range');
+
+	if (!rangeHeader) {
+		return new Response(object.body, {
+			status: 200,
+			headers: getHeaders(object, { type: 0 }),
 		});
+	}
 
-		let isR2ObjectBody = (object: R2Object | R2ObjectBody): object is R2ObjectBody => {
-			return 'body' in object;
+	const ranges = rangeHeader.replace(/bytes=/, '').split(',');
+	if (ranges.length === 1) {
+		const [start, end] = ranges[0].split('-').map(Number);
+		if (!end || start >= object.size || end >= object.size || (end !== 0 && start >= end)) {
+			return new Response('Range Not Satisfiable', { status: 416 });
 		}
+		return new Response(object.body, {
+			status: 206,
+			headers: getHeaders(object, { type: 1, start, end }),
+		});
+	}
 
-		if (object === null) {
-			return new Response('Not Found', { status: 404 });
-		} else if (!isR2ObjectBody(object)) {
-			return new Response("Precondition Failed", { status: 412 });
-		} else {
-			return new Response(object.body, {
-				status: object.range ? 206 : 200,
-				headers: {
-					'Content-Type': object.httpMetadata?.contentType ?? 'application/octet-stream',
-					// TODO: Content-Length, Content-Range
-
-					...(object.httpMetadata?.contentDisposition ? {
-						'Content-Disposition': object.httpMetadata.contentDisposition,
-					} : {}),
-					...(object.httpMetadata?.contentEncoding ? {
-						'Content-Encoding': object.httpMetadata.contentEncoding,
-					} : {}),
-					...(object.httpMetadata?.contentLanguage ? {
-						'Content-Language': object.httpMetadata.contentLanguage,
-					} : {}),
-					...(object.httpMetadata?.cacheControl ? {
-						'Cache-Control': object.httpMetadata.cacheControl,
-					} : {}),
-					...(object.httpMetadata?.cacheExpiry ? {
-						'Cache-Expiry': object.httpMetadata.cacheExpiry.toISOString(),
-					} : {}),
-				}
-			});
+	const r: [number, number][] = [];
+	for (const range of ranges) {
+		const [start, end] = range.split('-').map(Number);
+		if (!end || start >= object.size || end >= object.size || (end !== 0 && start >= end)) {
+			return new Response('Range Not Satisfiable', { status: 416 });
 		}
+		r.push([start, end]);
 	}
+
+	const boundary = `------------${crypto.randomUUID()}`;
+	const encode = new TextEncoder();
+	const rangeStream = new ReadableStream<Uint8Array>({
+		async start(controller) {
+			for (const [start, end] of r) {
+				const object = (await bucket.get(resource_path, {
```

Comment:
```suggestion
				// The privous 404 response has already handled, so it can't be null here.
				const object = (await bucket.get(resource_path, {
```